File: mvi/src/main/java/com/mdgd/mvi/states/AbstractEffect.java
package com.mdgd.mvi.states;

public class AbstractEffect<V> implements ScreenState<V> {
    private boolean isHandled = false;

    @Override
    public final void visit(V screen) {
        if (!isHandled) {
            handle(screen);
            isHandled = true;
        }
    }

    protected void handle(V screen) {

    }
}

File: mvi/src/main/java/com/mdgd/mvi/states/AbstractEffect.kt
package com.mdgd.mvi.states

abstract class AbstractEffect<T> : ScreenEffect<T> {
    var isHandled = false

    override fun visit(screen: T) {
        if (!isHandled) {
            handle(screen);
            isHandled = true
        }
    }

    abstract fun handle(screen: T);
}